linkerd/app/core/src/reject_with_no_identity.rs

```diff
@@ -0,0 +1,70 @@
+use super::transport::tls::accept;
+use crate::transport::io::BoxedIo;
+use crate::Error;
+use futures::future::{self, Either, Future, FutureResult};
+use futures::Poll;
+use linkerd2_proxy_core::listen::Accept;
+
+#[derive(Debug)]
+pub struct MissingIdentity;
+
+impl std::fmt::Display for MissingIdentity {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "identity is missing")
+    }
+}
+
+impl std::error::Error for MissingIdentity {}
+
+#[derive(Debug, Clone)]
+pub struct RejectWithNoIdentityLayer {
+    enabled: bool,
+}
+
+#[derive(Debug, Clone)]
+pub struct RejectWithNoIdentity<A> {
+    enabled: bool,
+    accept: A,
+}
+
+impl RejectWithNoIdentityLayer {
+    pub fn new(enabled: bool) -> Self {
+        Self { enabled }
```

Comment:
TIOLI, but we can decouple this middleware from `linkerd/app/core` and potentially make it more generally useful...

If we instead define a
```rust
pub trait Admit<T> {
  type Error: Into<Error>;
  fn admit(&mut self, target: &T) -> Result<(), Self::Error>;
}
```

Then (1) we can totally decouple this from "identity" -- the middleware is just a service and it doesn't even have to know about `Accept`, for that matter...

Then, up in app/core, we define an implementation that knows how to handle `(accept::Meta, BoxedIo)`. Here's a [sketch](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=401eb438c31a626c21f426b996ff7333).

Reply:
@olix0r Yes decoupling it makes sense. Seems like I am missing something but The one thing I did not quite get is how is the Admit a service? Or are you suggesting using the Admit in the same fashion we are using the  Detect and just passing it into a service that will either admit it and call the inner svc or not admit it and return the err.

Reply:
@zaharidichev The latter. Basically, modify this middleware to take an `Admit`. I'd then probably recommend moving this into a more generalized (or even dedicated) crate. Also.. we may already have a middleware that works like this. I'm not sure.